### ISSUE TITLE
fix: 🐛 Regression: Route guard never completes when app kicks off state in AppComponent

### ISSUE THREAD

**Issue by bjornharvold** [angular/angular#53247]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

With the introduction of Angular 17 and its... 

provideRouter(
   ROUTES,
   withEnabledBlockingInitialNavigation(),
)

...we have ourselves a bit of a conundrum and I am not sure how to handle it. This problem occurs when we navigate directly to a nested page. We do not start on AppComponent and let the user drill down manually. AppComponent contains nothing but a `<router-outlet></router-outlet>`.

We use Ngrx as our state management solution but our example removed Ngrx for simplicity to convey the problem.

Our App workflow goes like this:
- AppComponent kicks off some configuration tasks like 
   1. Checking which environment it's in 
   2. Which affiliate is loading  our app. 
   3. Remember me
...and dispatches the application state to get the ball rolling

In the example indicated below, imagine the /lazy route is the default lazy route to forward to upon application start. '/lazy' needs to know that state is in a good state before displaying the page. 

On 99% of our pages, we have CanActivate guards that dispatches state updates. Many of those guards rely on other state to be available before we dispatch more state changes. 

Example: User needs to be authenticated before we dispatch `loadTodoList()`. In Angular 16, we have in the guards, that require the user to be authenticated, a reactive check for whether the user is authenticated. Once the observable returns true, we execute the state update and return 'true' as guard results and the page displays.

In Angular 17, this no longer works.

In order to work around this new shortcoming, we had to deactivate all these guards and instead put all the "init logic" into the components; same for resolvers. Example: A resolver that would be responsible for setting the page title based on data that's returned from the server would also just hang the page because the state would never get into a ready-state. 

The issue here is similar to the documentation on ApplicationRef.isStable(): [angular.io/ApplicationRef.isStable()](https://angular.io/api/core/ApplicationRef#isstable-examples-and-caveats). We were trying to get certain state to kick off when isStable turned true, but it never did, because the route guard was waiting on a ready-state of its own. Catch-22.

This worked prior to Angular 17 so will call this a regression.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/~/github.com/bjornharvold/ngrx-issue-4149

### Please provide the exception or error you saw

```true
2 things can occur based on with withEnabledBlockingInitialNavigation():

1. When present, application never reaches lazy page and just hangs
2. When not present, guard immediately fails
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.0.3
Node: 20.10.0
Package Manager: yarn 1.22.21
OS: darwin arm64

Angular: 17.0.4
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.3
@angular-devkit/build-angular   17.0.3
@angular-devkit/core            17.0.3
@angular-devkit/schematics      17.0.3
@angular/cli                    17.0.3
@schematics/angular             17.0.3
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```

**Comment by atscott** [comment]
This fundamentally can't work if your route guards have something that depend on `AppComponent` being loaded. `withEnabledBlockingInitialNavigation` is supposed to block bootstrap of the application until navigation completes. Your navigation doesn't complete unless bootstrap already happened. Like you said, this is a catch 22. I don't know how or why this worked in v16 but the current behavior certainly seems correct.

At least in your example, the code that's in the `ngOnInit` of the `AppComponent` belongs in an `APP_INITIALIZER` instead. I know the example is simplified quite a bit, but maybe it's not quite showing the root cause. 

There _was_ a change to `ApplicationRef.isStable` behavior in v17 that does not emit stable until the initial render is complete, which includes the initial navigation. Previously, the stableness was only driven by ZoneJS. https://github.com/angular/angular/commit/28c68f709cdc930e12bac51a266e7bf790656034#diff-c3bc2f4d2923398d46897c1233ab19cec2611e44ea60592e7e0d5ed03747a897R847-R853

// cc @AndrewKushnir @alan-agius4

**Comment by atscott** [comment]
@bjornharvold Could you update your example a bit to show more clearly what's going on in your application? I don't think it's really showing what you described. As I mentioned in my previous comment, the example is showing a situation that I couldn't imagine ever worked. https://stackblitz.com/edit/stackblitz-starters-x8oytu is the example using v16 and does not work either (as expected).

**Comment by bjornharvold** [comment]
Hi @atscott,

Thank you for the quick response time. 

We've had this working for years with every version of Angular up until 17 so now I am perplexed. What exactly cannot work? Did standalone change something under the hood?

The use case that we are trying to solve is delayed guard function response while state changes in the background.

Example: Authentication
1. User clicks on a link from an external site that takes us deep into our app
2. She already has a user token in her session from earlier in a different browser tab
3. Remember Me needs to do its work to verify there already is a session (grab token from session, check w auth server etc)
4. Before we let her into that page, remember me needs to come back to us with a true / false. (if we ask the auth service if she's authenticated right away, it would fail every time)

Our state resides with Ngrx and we load up that state provider in our main.ts file. The first time we interact with that state is in AppComponent.

Now we have a few similar use cases of guards that wait for some state before dispatching other state changes. Most of the issues we've had is with state that init in AppComponent.

Instead of rebuilding our oauth2 "remember me" workflow as an example...

1. What is the Angular Team's best practice for bootstrapping initial state?
2. What is the Angular Team's best practice for writing a resolver or guard that requires them to wait on state to become available like the example I linked to above?

If we needed to move code out of AppComponent such as Remember Me functionality... and I do believe not all use cases are applicable to use with APP_INITIALIZER, could we chain guards such that the the root app route has a guard that dispatches Remember Me and other similar things. The lazy nested route that requires the user to be authenticated has a guard that waits on the remember me state to be initialized before checking if the user is authenticated.

Trying to find a beautiful working way to deal with the workflow of "app-init and state that depends on state" and how to properly utilize guards and resolvers so they never end up in a catch-22 situation.

🍺

**Comment by bjornharvold** [comment]
Hi @atscott,

Going to close this ticket. If you are saying "This is not how it's supposed to work and should've never worked" and I show a demo in Angular 16 with NgModules where it does work, we'd both agree that it is not possible in Angular 17 (with the standalone way of doings things at least) anyway and not a bug and you'd maybe recommend another way of doing things. So will save us both some time.

Parting thought:
Could be interesting to create some documentation on Angular Best Practices for advanced applications that would include:

1. How to properly init an app to avoid these catch-22s that can occur when a bunch of dependent state all wants to start up at once but have to wait while their dependencies initialize first.
2. How to never end up with route guards that block indefinitely because app is not stable, state is not ready or similar

Thank you for your time 🍺

**Comment by atscott** [comment]
> If you are saying "This is not how it's supposed to work and should've never worked"

That’s really not what I’m saying. The reproduction you have in this issue report did not work in v17 or v16. I was simply trying to understand what your actually state was because the reproduction couldn’t have been it. 

If you’re able to pin down a minimal reproduction that effectively demonstrates the behavior that worked for your application in v16 but no longer works in v17, please do open a new issue.
